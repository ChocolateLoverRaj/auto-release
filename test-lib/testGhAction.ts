import { spawn } from 'child_process'
import streamToString from 'stream-to-string'
import { once } from 'events'
import { EOL } from 'os'
import { join } from 'path'
import { writeFile } from 'jsonfile'
import defaults from 'object.defaults'
import getPort from 'get-port'
import express from 'express'
import { createServer } from 'http'

interface Result {
  outputs: Record<string, string>
  stdout: string
}

interface Commit {
  message: string
}

interface PullRequest {
  commits: Commit[]
}

interface Repo {
  pullRequests: Record<number, PullRequest>
  token?: string
}

interface PrEvent {
  number: number
}

interface Options {
  inputs: Record<string, string>
  event: PrEvent
  repo: Repo
  env: Record<string, string>
}

const testGhAction = async (file: string, partialOptions: Partial<Options> = {}): Promise<Result> => {
  const options: Options = defaults(partialOptions, { inputs: [], event: { number: 1 }, repo: { pullRequests: {} } })
  const { event, inputs, repo } = options
  const eventFile = join(__dirname, '../test-tmp/event.json')
  const port = await getPort()

  const server = express()
  const fakeGhServer = createServer(server).listen(port)

  server.get('/repos/ChocolateLoverRaj/test-semver/pulls/:number/commits', (req, res) => {
    const n = parseInt(req.params.number)
    const pr = repo.pullRequests[n]
    if (pr === undefined) {
      res.sendStatus(404)
      return
    }
    res.json(pr.commits.map(commit => ({ commit })))
  })

  await Promise.all([
    writeFile(eventFile, {
      pull_request: {
        commits: repo.pullRequests[event.number].commits.length,
        number: event.number
      },
      repo: {
        name: 'test-semver',
        owner: {
          login: 'ChocolateLoverRaj'
        }
      }
    }, { spaces: 2 }),
    once(fakeGhServer, 'listening')
  ])

  const c = spawn('node', [file], {
    env: {
      ...process.env,
      GITHUB_EVENT_PATH: eventFile,
      GITHUB_API_URL: `http://localhost:${port}`,
      ...Object.fromEntries(Object.entries(inputs).map(([k, v]) => [`INPUT_${k.toUpperCase()}`, v])),
      ...options.env
    }
  })
  const stdoutStrPromise = streamToString(c.stdout)
  const stderrStr = streamToString(c.stderr)
  const [code] = await once(c, 'exit')
  fakeGhServer.close()
  if (code !== 0) {
    throw new Error(
      `Process exited with code ${code as number}\n` +
      'Stderr is below:\n' +
      await stderrStr
    )
  }
  const setOutputStr = '::set-output name='
  const stdoutStr = (await stdoutStrPromise)
  return {
    outputs: Object.fromEntries(stdoutStr
      .split(EOL)
      .map(str => {
        if (str.startsWith(setOutputStr)) {
          const keyValue = str.slice(setOutputStr.length).split('::')
          if (keyValue.length !== 2) throw new Error('Bad set output syntax')
          return keyValue
        }
        return undefined
      })
      .filter(v => v !== undefined) as any
    ),
    stdout: stdoutStr
  }
}

export default testGhAction
