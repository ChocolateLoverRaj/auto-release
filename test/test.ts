import testGhAction from '../test-lib/testGhAction'
import { join } from 'path'

test('works', async () => {
  const { stdout } = await testGhAction(join(__dirname, '../dist/index.js'), {
    event: {
      number: 1
    },
    repo: {
      pullRequests: {
        1: {
          commits: [{
            message: 'Feature: add triangles'
          }, {
            message: 'Breaking: replaced triangle with polygon'
          }, {
            message: 'Fix: polygons have minimum of 3 sides'
          }, {
            message: 'Chore: updated GitHub action script'
          }]
        }
      },
      token: 'token'
    },
    env: {
      GITHUB_TOKEN: 'token'
    }
  })
  console.log(stdout)
})

test('invalid commit message', async () => {
  await expect(testGhAction(join(__dirname, '../dist/index.js'), {
    event: {
      number: 1
    },
    repo: {
      pullRequests: {
        1: {
          commits: [{ message: 'made it better' }]
        }
      },
      token: 'token'
    },
    env: {
      GITHUB_TOKEN: 'token'
    }
  })).rejects.toMatchSnapshot()
})
