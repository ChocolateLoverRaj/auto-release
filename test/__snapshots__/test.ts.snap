// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`invalid commit message 1`] = `
[Error: Process exited with code 1
Stderr is below:
Error: Invalid Commit Message: made it better. parse-commit-message: expect \`commit\` to follow:
<type>[optional scope]: <description>

[optional body]

[optional footer]
    at C:\\Users\\rajas\\Documents\\GitHub\\pr-increment-preview\\dist\\index.js:34:19
    at Array.map (<anonymous>)
    at C:\\Users\\rajas\\Documents\\GitHub\\pr-increment-preview\\dist\\index.js:27:51
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
]
`;
